### PR TITLE
Deprecate manually adding configurations to configuration container

### PR DESCRIPTION
**Pull request by jvandort**
Detached configurations and copied configurations, which are also detached configurations, should not be added to the configuration container. As such, we deprecate calling any add method to the container, restricting the user to only add configurations using the standard factory methods.

Furthermore, we modify the roles given to copied configurations. Previously, we used the roles of the original configuration. However, when a configuration is copied, all of its dependencies, even from the configurations it extends from, are accumulated and added to the detached configuration. This means that the copied configuration will have dependencies and will be resolvable. As such, it must be a resolvable dependency scope.

Fixes: #24895 

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc [hunk 1]
```diff
@@ -206,6 +142,78 @@ The existing method does not behave as specified by its javadoc, and we do not p
 
 Potential migrations include using `Object.equals(Object)` directly, or comparing the fields of dependencies manually.
 
+[[changes_8.10]]
```
Comment (jvandort) [inline platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:145]: There were a number of deprecations added in 8.11 that were erroneously added to the 8.10 section. They were moved up to 8.11 in this PR. That is why the diff looks like this

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test this

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationContainer.java [hunk 2]
```diff
@@ -138,6 +140,50 @@ protected UnknownDomainObjectException createNotFoundException(String name) {
...
+            .withAdvice("Use a factory method instead.")
+            .willBeRemovedInGradle9()
+            .withUpgradeGuideSection(8, "adding_to_configuration_container")
+            .nagUser();
+        super.addAllLater(provider);
+    }
+
+    private void addInternal(Configuration configuration) {
+        super.add(configuration);
+    }
```
Comment (tresat) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationContainer.java:185]: 💅 Seems unnecessary, just call `super.add` from the one usage of this instead.

**Review by tresat — APPROVED** [review]

**Pull request merged**